### PR TITLE
*: milisecond test and infoschema fix

### LINKED ISSUE #24812: `TestStalenessTransactionSchemaVer` is failed
## Bug Report

```
[2021-05-21T03:46:36.263Z] ----------------------------------------------------------------------
[2021-05-21T03:46:36.263Z] FAIL: stale_txn_test.go:297: testStaleTxnSerialSuite.TestStalenessTransactionSchemaVer
[2021-05-21T03:46:36.263Z] 
[2021-05-21T03:46:36.263Z] stale_txn_test.go:325:
[2021-05-21T03:46:36.263Z]     // got an old infoSchema
[2021-05-21T03:46:36.263Z]     c.Assert(schemaVer6, Equals, schemaVer4)
[2021-05-21T03:46:36.263Z] ... obtained int64 = 34
[2021-05-21T03:46:36.263Z] ... expected int64 = 31
[2021-05-21T03:46:36.263Z] 
```

### PR DESCRIPTION
**Pull request by xhebox**
### What problem does this PR solve?

Issue Number: close #24812 

Problem Summary: Failed because of time prec. While fixing the problem, I found that when `commit` is executed, `TxnCtx.InfoSchema` is not refreshed, it is lazily refreshed. Then `sessionVars.GetInfoSchema()` will return the old infoschema after `commit`.

Later commits will hopefully remove the usage of `dom.InfoSchema()` and refine `TxnCtx`. But for now, let us merge this PR first.

### What is changed and how it works?

What's Changed:

1. Add check `vars.TxnCtx != nil && vars.TxnCtx.IsExplicit` for `GetInfoSchema()`. Now it will not return the old infoschema after `commit`, but will still return nil.
2.  Add `domain.GetDomain(sctx).InfoSchema()` for `GetInfoSchema()` to return the latest commit after `commit`. But there is a circular dependency loop, it is moved from a getter of `sessionVars` to a method of `sessionctx.Context`.
3. Remove `builder.Request.SchemaVar = is.SchemaMetaVersion()` in `SetFromSessionVars` and add `builder.Request.SchemaVar = is.SchemaMetaVersion()` in `SetFromInfoSchema`. Several `SetFromInfoSchema` was added to ensure infoschema is set correctly.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- No release note

### CODE REVIEW

File: executor/table_reader.go [hunk 1]
```diff
@@ -228,13 +228,10 @@ func (e *TableReaderExecutor) buildResp(ctx context.Context, ranges []*ranger.Ra
...
 		SetStreaming(e.streaming).
 		SetFromSessionVars(e.ctx.GetSessionVars()).
+		SetFromInfoSchema(e.ctx.GetInfoSchema().(infoschema.InfoSchema)).
 		SetMemTracker(e.memTracker).
 		SetStoreType(e.storeType).
 		SetAllowBatchCop(e.batchCop)
-	// infoschema maybe null for tests
-	if is, ok := e.ctx.GetSessionVars().GetInfoSchema().(infoschema.InfoSchema); ok {
-		reqBuilder.SetFromInfoSchema(is)
-	}
```
Comment (JmPotato) [inline executor/table_reader.go:237]: I think this part of logic is not the same as before? `nil` interface should be handled.

Comment (xhebox) [reply executor/table_reader.go:237]: I thought `GetInfoSchema` will always return something. I did not expect that `domain.InfoSchema()` will return nil.

Comment (xhebox) [reply executor/table_reader.go:237]: Fixed. Now request builder will check it by `is, ok := emptyInterce.(InfoSchema)`.

**Comment by xhebox** [comment]
OK, a mistake. Looks like `IsExplicit` can not always be updated.... It is more like a one-shot variable. I've switched to `GetStatusFlag()`.
/run-all-tests

**Comment by xhebox** [comment]
OK.. Got another panic and fixed. So planner relies on that snapshot infoschema for testing, and I must provide a partial mock..

**Comment by xhebox** [comment]
Why trigger all tests again and again? From the log, the error is reported by `tidb-test/sqllogic_test/main.go`

**Comment by purelind** [comment]
> Why trigger all tests again and again? From the log, the error is reported by `tidb-test/sqllogic_test/main.go`

CI just upgrade go version from 1.13 to 1.16.4. Here we want to check again

**Review by JmPotato — APPROVED** [review]

**Review by Yisaer — APPROVED** [review]
/lgtm

File: sessionctx/context.go [hunk 2]
```diff
@@ -28,6 +28,14 @@ import (
 	"github.com/pingcap/tipb/go-binlog"
 )
 
+// InfoschemaMetaVersion is a workaround. Due to circular dependency,
+// can not return the complete interface. But SchemaMetaVersion is widely used for logging.
+// So we give a convenience for that.
+// FIXME: remove this interface
+type InfoschemaMetaVersion interface {
+	SchemaMetaVersion() int64
```
Comment (AilinKid) [inline sessionctx/context.go:36]: For interface used as a tool to avoid circular, we can put it in the util package.
But for this, it's ok.

**Pull request merged**